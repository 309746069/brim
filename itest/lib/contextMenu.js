/* @flow */
import type {Space} from "../../src/js/lib/Space"
import Field from "../../src/js/models/Field"
import Log from "../../src/js/models/Log"
import menu from "../../src/js/electron/menu"

export default function contextMenu(
  app: *,
  program: string,
  columns: string[],
  space: Space
) {
  return {
    program(value: string) {
      program = value
      return this
    },

    click(label: string, field: Field, log: Log) {
      let item = menu
        .fieldContextMenu(program, columns, space)(
          field.toBrimField(),
          log,
          false
        )
        .find((item) => item.label && item.label === label)

      if (item && item.click) {
        item.click(item, {webContents: app.webContents})
      } else {
        throw new Error("Could not find '" + label + "' in context menu")
      }

      return this
    }
  }
}
