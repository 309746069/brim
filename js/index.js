import React from "react"
import ReactDOM from "react-dom"
import {AppContainer} from "react-hot-loader"
import {Provider} from "react-redux"
import initStore from "./initStore"
import {HashRouter} from "react-router-dom"
import XApp from "./connectors/XApp"
const store = initStore()

ReactDOM.render(
  <AppContainer>
    <Provider store={store}>
      <HashRouter>
        <XApp />
      </HashRouter>
    </Provider>
  </AppContainer>,
  document.getElementById("app-root")
)
