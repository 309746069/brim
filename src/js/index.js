import React from "react"
import ReactDOM from "react-dom"
import {Provider} from "react-redux"
import initStore from "./initStore"
import {HashRouter} from "react-router-dom"
import XApp from "./connectors/XApp"
const store = initStore()

ReactDOM.render(
  <Provider store={store}>
    <HashRouter>
      <XApp />
    </HashRouter>
  </Provider>,
  document.getElementById("app-root")
)
