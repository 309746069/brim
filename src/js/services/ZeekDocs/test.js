/* @flow */
import ZeekDocs from "./"

test("describe conn uid", () => {
  let desc = ZeekDocs.describe("conn", "uid")

  expect(desc).toBe("A unique identifier of the connection.")
})
