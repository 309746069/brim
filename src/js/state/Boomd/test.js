/* @flow */
import Boomd from "./"
import initTestStore from "../../test/initTestStore"

test("setting use analytics cache", () => {
  const store = initTestStore()
  store.dispatch(Boomd.useCache(true))
  expect(Boomd.usingCache(store.getState())).toEqual(true)
  store.dispatch(Boomd.useCache(false))
  expect(Boomd.usingCache(store.getState())).toEqual(false)
})

test("setting use index", () => {
  const store = initTestStore()
  store.dispatch(Boomd.useIndex(true))
  expect(Boomd.usingIndex(store.getState())).toEqual(true)
  store.dispatch(Boomd.useIndex(false))
  expect(Boomd.usingIndex(store.getState())).toEqual(false)
})
