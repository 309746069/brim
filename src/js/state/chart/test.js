/* @flow */

import chart from "./"
import initTestStore from "../../test/initTestStore"

let store
beforeEach(() => {
  store = initTestStore()
})

let records = [
  [{name: "id", type: "string", value: "1"}],
  [{name: "id", type: "string", value: "2"}]
]

test("chart records append", () => {
  let state = store.dispatchAll([
    chart.appendRecords([records[0]]),
    chart.appendRecords([records[1]])
  ])

  expect(chart.getRecords(state)).toEqual(records)
})

test("chart records status", () => {
  let state = store.dispatchAll([chart.setStatus("SUCCESS")])

  expect(chart.getStatus(state)).toBe("SUCCESS")
})

test("chart records clear", () => {
  let state = store.dispatchAll([
    chart.appendRecords(records),
    chart.setStatus("SUCCESS"),
    chart.clear()
  ])

  expect(chart.getStatus(state)).toBe("INIT")
  expect(chart.getRecords(state)).toEqual([])
})
