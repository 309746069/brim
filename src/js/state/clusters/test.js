/* @flow */

import {addCluster, removeCluster, setCluster, setClusterError} from "./actions"
import {
  getCurrentCluster,
  getClusterMessage,
  getSavedClusters
} from "./selectors"
import initTestStore from "../../test/initTestStore"

let store
beforeEach(() => {
  store = initTestStore()
})

let cluster = {
  host: "boom.com",
  port: "9867",
  username: "kerr",
  password: "123"
}

test("addCluster", () => {
  let state = store.dispatchAll([addCluster(cluster)])

  expect(getSavedClusters(state)).toEqual([cluster])
})

test("addCluster when it already exists", () => {
  let state = store.dispatchAll([addCluster(cluster), addCluster(cluster)])

  expect(getSavedClusters(state)).toEqual([cluster])
})

test("removeCluster", () => {
  let state = store.dispatchAll([addCluster(cluster), removeCluster(cluster)])

  expect(getSavedClusters(state)).toEqual([])
})

test("setCluster", () => {
  let state = store.dispatchAll([setCluster(cluster)])

  expect(getCurrentCluster(state)).toEqual(cluster)
})

test("setClusterMessage to message", () => {
  let state = store.dispatchAll([setClusterError("no spaces")])

  expect(getClusterMessage(state)).toEqual("no spaces")
})

test("setClusterMessage to empty", () => {
  let state = store.dispatchAll([setClusterError("")])

  expect(getClusterMessage(state)).toEqual("")
})
