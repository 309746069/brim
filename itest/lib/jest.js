/* @flow */

import {Application} from "spectron"

import {selectors} from "../../src/js/test/integration"

export const TestTimeout = 60000

// This is trying to make Flow happy for passing in the done function from
// Jest.
export type Done = {
  (string | Error | void): void,
  fail?: (string | Error | void) => void
}

export const handleError = async (
  app: Application,
  initialError: Error,
  done: Done
) => {
  let realError = undefined
  let notificationError = undefined
  console.log(`handleError: Test hit exception: ${initialError.message}`)
  console.log("handleError: Looking for any desktop app notifications")
  try {
    notificationError = await app.client.getHTML(selectors.notification, false)
  } catch (e) {
    notificationError = undefined
  }
  if (notificationError) {
    realError = new Error(
      "App notification '" +
        notificationError +
        "' (initial error: '" +
        initialError +
        "'"
    )
  } else {
    console.log("handleError: desktop app notification not found")
    realError = initialError
  }
  done.fail && done.fail(realError)
}
