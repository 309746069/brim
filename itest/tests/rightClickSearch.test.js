/* @flow */

import {basename} from "path"

import {dataSets, selectors} from "../../src/js/test/integration"
import {
  getSearchText,
  logIn,
  newAppInstance,
  resetState,
  searchDisplay,
  setSpan,
  startApp,
  startSearch,
  waitForSearch,
  writeSearch
} from "../lib/app.js"
import {handleError, stdTest} from "../lib/jest.js"
import contextMenu from "../lib/appIpc"
import fixtures from "../../src/js/test/fixtures"
import mockSpace from "../../src/js/test/mockSpace"

describe("Test search mods via right-clicks", () => {
  let app
  let testIdx = 0
  let menu

  beforeEach(() => {
    app = newAppInstance(basename(__filename), ++testIdx)
    menu = contextMenu(app, "*", ["ts", "uid", "_path"], mockSpace())
    return startApp(app)
  })

  afterEach(async () => {
    if (app && app.isRunning()) {
      await resetState(app)
      return await app.stop()
    }
  })

  stdTest("Include / Exclude this value works", (done) => {
    let includeExcludeFlow = async () => {
      await logIn(app)

      let field = dataSets.corelight.rightClickSearch.includeValue
      let log = fixtures.log("conn1")
      menu.click("Filter = value", field, log)
      menu.click("Filter != value", fixtures.field("conn"), log)

      // The result order is deterministic because clicked a uid and then
      // removed its conn log entry.
      return searchDisplay(app)
    }
    includeExcludeFlow()
      .then((searchResults) => {
        expect(searchResults).toMatchSnapshot()
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest("Use as start/end time works", (done) => {
    let startEndFlow = async () => {
      await logIn(app)
      // The sort proc is used here to ensure that the two tuples that appear
      // with the same timestamp are deterministically sorted.  If PROD-647
      // is fixed, the results will still be deterministic because, although
      // two points with the same ts appear, they will be sorted by uid as
      // well.
      let program = "_path=conn | sort -r ts, uid"
      await writeSearch(app, program)
      await startSearch(app)
      await waitForSearch(app)

      let {startTime, endTime} = dataSets.corelight.rightClickSearch
      let log = fixtures.log("conn1")

      menu
        .program(program)
        .click('Use as "start" time', startTime, log)
        .click('Use as "end" time', endTime, log)

      return searchDisplay(app)
    }
    startEndFlow()
      .then((searchResults) => {
        expect(searchResults).toMatchSnapshot()
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest("New Search works", (done) => {
    let newSearchFlow = async () => {
      await logIn(app)
      let uid = dataSets.corelight.rightClickSearch.newSearchSetup
      let path = fixtures.field("weird")
      let log = fixtures.log("weird1")
      menu
        .click("Filter = value", uid, log)
        .click("New search with this value", path, log)
      // The result order is deterministic because all the points rendered have
      // different ts values.
      return searchDisplay(app)
    }
    newSearchFlow()
      .then((searchResults) => {
        expect(searchResults).toMatchSnapshot()
      })
      // This section verifies the previous search was cleared in favor of
      // the new search "weird".
      .then(() => getSearchText(app))
      .then((searchText) => {
        expect(searchText).toBe('"weird"')
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest("Count by / Pivot to logs works", (done) => {
    let pivotToLogsFlow = async () => {
      await logIn(app)
      // If we want to "Count by" _path, we *must* select a conn cell,
      // because conn logs contain the _path of the connection. Searching for
      // "dns" will find either the first conn log tuple of a dns connection,
      // or the first dns log tuple of the same.
      menu
        .click("Count by field", fixtures.field("conn"), fixtures.log("conn1"))
        .program(await getSearchText(app))
        .click("Pivot to logs", fixtures.field("dhcp"), fixtures.log("dhcp1"))

      // The result order is deterministic because all the points rendered have
      // different ts values.
      return searchDisplay(app)
    }
    pivotToLogsFlow().then((searchResults) => {
      expect(searchResults).toMatchSnapshot()
      done()
    })
  })

  stdTest("conn for www.mybusinessdoc.com is found via correlation", (done) => {
    logIn(app)
      .then(() => setSpan(app, dataSets.corelight.logDetails.span))
      .then(() => writeSearch(app, dataSets.corelight.logDetails.initialSearch))
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      // Search is deterministic because all tuples have different ts
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => {
        let uid = dataSets.corelight.logDetails.getDetailsFrom
        menu
          .program(dataSets.corelight.logDetails.initialSearch)
          .click("Open details", uid, fixtures.log("myBusinessDocHttp"))
      })
      .then(() =>
        Promise.all([
          app.client
            .waitForVisible(selectors.correlationPanel.tsLabel)
            .then(() => app.client.getText(selectors.correlationPanel.tsLabel)),
          app.client
            .waitForVisible(selectors.correlationPanel.pathTag)
            .then(() => app.client.getText(selectors.correlationPanel.pathTag)),
          app.client
            .waitForVisible(selectors.correlationPanel.duration)
            .then(() => app.client.getText(selectors.correlationPanel.duration))
            .then((result) => [result])
        ])
      )
      .then((correlationData) => {
        expect(correlationData).toMatchSnapshot()
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })
})
