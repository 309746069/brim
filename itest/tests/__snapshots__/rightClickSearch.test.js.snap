// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test search mods via right-clicks Include / Exclude this value works 1`] = `
Array [
  "ts",
  "_path",
  "uid",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "proto",
  "trans_id",
  "rtt",
  "query",
  "qclass",
  "qclass_name",
  "qtype",
  "qtype_name",
  "rcode",
  "rcode_name",
  "AA",
  "TC",
  "RD",
  "RA",
  "Z",
  "answers",
  "TTLs",
  "rejected",
  "04/13/15",
  "09:34",
  "44.7270",
  "dns",
  "CgVEQNkcytYYMtmE5",
  "192.168.0.51",
  "5353",
  "224.0.0.251",
  "5353",
  "udp",
  "0",
  "-",
  "_workstation._tcp.local",
  "-",
  "-",
  "-",
  "-",
  "0",
  "NOERROR",
  "T",
  "F",
  "F",
  "F",
  "0",
  "homer-xubuntu [ec:f4:bb:4f:b2:45]._workstation._tcp.local,homer-xubuntu.local,TXT 0 ,_workstation._tcp.local,192.168.0.51,homer-xubuntu.local,fe80::eef4:bbff:fe4f:b245,homer-xubuntu.local",
  "0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000",
  "F",
]
`;

exports[`Test search mods via right-clicks New Search works 1`] = `
Array [
  "ts",
  "_path",
  "uid",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "name",
  "addl",
  "notice",
  "peer",
  "04/13/15",
  "09:32",
  "45.1400",
  "weird",
  "CZehjU3zE0mxEXiC41",
  "192.168.0.53",
  "3683",
  "192.168.0.1",
  "8080",
  "unknown_HTTP_method",
  "UNSUBSCRIBE",
  "F",
  "bro",
  "04/13/15",
  "09:32",
  "45.1220",
  "weird",
  "Ctd0Gv1XC3Z0UV1iX",
  "192.168.0.53",
  "3682",
  "192.168.0.1",
  "8080",
  "unknown_HTTP_method",
  "UNSUBSCRIBE",
  "F",
  "bro",
  "04/13/15",
  "09:31",
  "51.9870",
  "weird",
  "-",
  "-",
  "-",
  "-",
  "-",
  "unknown_protocol_2",
  "-",
  "F",
  "bro",
  "04/13/15",
  "09:31",
  "27.8210",
  "weird",
  "CbtgYm2YYGmfYsyaC4",
  "192.168.0.54",
  "53147",
  "216.58.209.134",
  "80",
  "inflate_failed",
  "-",
  "F",
  "bro",
]
`;

exports[`Test search mods via right-clicks Pivot to logs works 1`] = `
Array [
  "ts",
  "_path",
  "uids",
  "client_addr",
  "server_addr",
  "mac",
  "host_name",
  "client_fqdn",
  "domain",
  "requested_addr",
  "assigned_addr",
  "lease_time",
  "client_message",
  "server_message",
  "msg_types",
  "duration",
  "04/13/15",
  "09:31",
  "26.5080",
  "dhcp",
  "CyZar62ctcnhulIaNa,Ca2wwF3Suk40Z1cvB1",
  "192.168.0.54",
  "-",
  "ec:f4:bb:4f:b0:96",
  "Dell-Dator32",
  "-",
  "pwned.se",
  "-",
  "-",
  "-",
  "-",
  "-",
  "INFORM,ACK",
  "0.024792",
  "04/13/15",
  "09:21",
  "48.1440",
  "dhcp",
  "CT4qoR3HvydUH88jL5,C55csu2Gcl4uqjpq3h",
  "192.168.0.54",
  "-",
  "ec:f4:bb:4f:b0:96",
  "Dell-Dator32",
  "-",
  "pwned.se",
  "-",
  "-",
  "-",
  "-",
  "-",
  "INFORM,ACK",
  "0.011216",
  "04/13/15",
  "09:11",
  "45.2980",
  "dhcp",
  "CsCHyE2ASmdH86SpPj,CobQkE2ubFvnevozP",
  "192.168.0.54",
  "-",
  "ec:f4:bb:4f:b0:96",
  "Dell-Dator32",
  "-",
  "pwned.se",
  "-",
  "-",
  "-",
  "-",
  "-",
  "INFORM,ACK",
  "0.016169",
]
`;

exports[`Test search mods via right-clicks Use as start/end time works 1`] = `
Array [
  "ts",
  "_path",
  "uid",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "proto",
  "service",
  "duration",
  "orig_bytes",
  "resp_bytes",
  "conn_state",
  "local_orig",
  "local_resp",
  "missed_bytes",
  "history",
  "orig_pkts",
  "orig_ip_bytes",
  "resp_pkts",
  "resp_ip_bytes",
  "tunnel_parents",
  "04/13/15",
  "09:34",
  "13.2420",
  "conn",
  "CZusuU27VRdyfg5thj",
  "192.168.0.51",
  "138",
  "192.168.0.255",
  "138",
  "udp",
  "-",
  "0.000031",
  "474",
  "0",
  "S0",
  "T",
  "T",
  "0",
  "D",
  "2",
  "530",
  "0",
  "0",
  "-",
  "04/13/15",
  "09:33",
  "03.2450",
  "conn",
  "Ct5sPR3IvG5vk8kg5a",
  "192.168.0.51",
  "5353",
  "224.0.0.251",
  "5353",
  "udp",
  "dns",
  "-",
  "-",
  "-",
  "S0",
  "T",
  "F",
  "0",
  "D",
  "1",
  "100",
  "0",
  "0",
  "-",
  "04/13/15",
  "09:33",
  "03.2450",
  "conn",
  "Clu1U4356BfPVhdLQ1",
  "192.168.0.2",
  "5353",
  "224.0.0.251",
  "5353",
  "udp",
  "dns",
  "-",
  "-",
  "-",
  "S0",
  "T",
  "F",
  "0",
  "D",
  "1",
  "364",
  "0",
  "0",
  "-",
  "04/13/15",
  "09:32",
  "45.1380",
  "conn",
  "CZehjU3zE0mxEXiC41",
  "192.168.0.53",
  "3683",
  "192.168.0.1",
  "8080",
  "tcp",
  "http",
  "0.004476",
  "241",
  "19",
  "S3",
  "T",
  "T",
  "0",
  "ShADadf",
  "4",
  "409",
  "4",
  "183",
  "-",
]
`;

exports[`Test search mods via right-clicks conn for www.mybusinessdoc.com is found via correlation 1`] = `
Array [
  "ts",
  "_path",
  "uid",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "trans_depth",
  "method",
  "host",
  "uri",
  "referrer",
  "version",
  "user_agent",
  "request_body_len",
  "response_body_len",
  "status_code",
  "status_msg",
  "info_code",
  "info_msg",
  "tags",
  "username",
  "password",
  "proxied",
  "orig_fuids",
  "orig_filenames",
  "orig_mime_types",
  "resp_fuids",
  "resp_filenames",
  "resp_mime_types",
  "04/07/15",
  "13:34",
  "44.2440",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "1",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=3271&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "318,883",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "FGx5ts2iCMfZSUgO8c",
  "c87ed3c.gif",
  "application/x-dosexec",
  "04/07/15",
  "13:34",
  "46.7830",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "2",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=5292&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "192,512",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "FuFeW33aTSTxXihCql",
  "551d88323f7e.gif",
  "application/x-dosexec",
  "04/07/15",
  "13:34",
  "48.1830",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "3",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=9813&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "1,536",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "F2rjhJZAjwGdIvTM8",
  "f7.gif",
  "application/x-dosexec",
]
`;

exports[`Test search mods via right-clicks conn for www.mybusinessdoc.com is found via correlation 2`] = `
Array [
  Array [
    "13:34:43.772",
    "13:34:44.244",
    "13:34:44.563",
    "13:34:46.783",
    "13:34:47.442",
    "13:34:48.183",
    "13:34:48.373",
  ],
  Array [
    "conn",
    "http",
    "files",
    "http",
    "files",
    "http",
    "files",
  ],
  Array [
    "Duration: 4.601s",
  ],
]
`;
